### DISCUSSION TITLE
Help in --tracks

### DISCUSSION THREAD

**Discussion by chickeneps** [keirf/greaseweazle#203] [General]
OK, to me I can't put things together well. I'm trying to read a Ensoniq DS/DD floppy that is 80/2/10x512, which means 80 tracks, 2 heads, 10 sectors per track and 512 bytes per sector. I want to end up with a 819200 byte image file that reflects the raw data on that floppy.

How does that translate to a gw read command? I see the example but it doesn't say anything about sector amount per track.

And, I might was well ask now; I'm going to want to reverse the process using write, and the easy part is that it'll be the same command except I'll use 'write' and not 'read'. However, does this have to be done on a already-formatted floppy? If so, fine - but further... is there any way in GW to format a floppy - that is, put the address marks in so a 'write' command can succeed? Or does GW format the floppy as it writes the image file?

**Comment by keirf** [comment]
The easiest thing is for me to add Ensoniq support to gw. I have just done this, so please try reinstalling the tools, something like:
```
python3 -m pip install git+https://github.com/keirf/greaseweazle@issue-203
```

Now you should be able to read an Ensoniq DD disk:
```
gw read --format ensoniq.800 my.img
```

Writing back will be as simple as changing read to write in the command line.

Let me know how it goes. I will commit this change to master for next proper release if it's working okay for you.

**Comment by chickeneps** [comment]
Thank you. I'm assuming that you just did DS/DD (ensoniq.800), can you do DS/HD (which I assume would be ensoniq.1600)?

Since I have a number of formats to deal with (Akai MPC60, AkaiS950 (DD and HD), Akai S1000/3000 (DD and HD), Emu Emax, Emu ESi/E4 (DD and HD), there's a couple more), and I believe you know of them (re: host= and interface= in FF), can you make it a project and include all of those?

My overarching project here is to create a future-proof sampler proprietary floppy reading/writing solution, as I look at my current XP32 computer with the floppy drive an one day it's gong to die.

Also, again while I have you, a couple things:

1) Ensoniq Mirage, since FF is HFE only on Mirage, same with GW? And thus is write-only, you can't read Mirage?

2) When you are "including support" like the above, you are including the specifics of that floppy layout since GW is starting "flat-footed" and it has to know ahead of time how to read the floppy? (I read that it just is so that GW doesn't have auto-format detect like OmniFlop. That's no big deal since I almost always know what the floppy format is.)

3) Similarly, any HFE file doesn't need any additional write support, since it's embedded in the file, correct?

Again, as always, thank you!

**Reply by keirf** [reply]
I did `ensoniq.1600` too. You can see the list if you `gw read -h`

**Reply by keirf** [reply]
1. Mirage is tougher. What you can do right now (and for most synth formats actually) is read to SCP, and then load that into HxC tools and save as IMG. To convert other way "Load Raw Image" in HxC tools, select predefined disk layout, and load your IMG, then export as HFE, and "gw write" that HFE.

**Reply by keirf** [reply]
2. Yes that's right GW will not work out the format of the IMG given the raw dump, it has to be told, and currently it has to be built into GW's Python code: it's not configurable on command line for example. And for write, you would generally need a hint about how to interpret the IMG anyway as sector headers are missing, for example.

**Reply by keirf** [reply]
3. Correct, it is raw track data which can be raw written to disk. Note that there is no verification of what's written. Unlike IMG, where GW will verify your write for you.

**Reply by keirf** [reply]
I don't know much about Emu, but Akai is easy enough, there are 800 and 1600kB (DD, HD) variants I will add to greaseweazle.

**Reply by keirf** [reply]
Okay if you `python3 -m pip install git+https://github.com/keirf/greaseweazle@issue-203` again now you should see the akai formats

**Comment by cbrunschen** [comment]
One thing to note with the Ensoniq formats is that they start sector numbering at zero (0), not one (1) as usual.

**Reply by keirf** [reply]
Yes that's handled by the new format specifiers.

**Comment by chickeneps** [comment]
Thank you much - this is huge... after decades of using DOS PC's, Windows PC's, pay money for developers to access floppy controllers, hours spent typing out why Mac's can't access Ensoniq floppies - finally a Mac accessed (r/w) an Ensoniq floppy here! Historic.

I read and wrote both DD and HD variants, works great. I'll start testing the other stuff shortly. I'll give you the Emu info, I think it's just like Ensoniq - DD=10 sectors HD=20 sectors per track.

Thanks again!

**Reply by keirf** [reply]
ensoniq and akai formats are in 1.0 release!
